File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.py
#User function Template for python3

class Solution:
    #Function to count the number of possible triangles.
    def countTriangles(self, arr):
        # code here
        arr.sort()
        n,ans=len(arr),0
        for i in range(n-1,1,-1):
            l,h=0,i-1
            while l<h:
                if arr[l]+arr[h]>arr[i]:
                    ans+=(h-l)
                    h-=1
                else:
                    l+=1
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.countTriangles(arr))

        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Count the number of possible triangles/count-the-number-of-possible-triangles.java
class Solution {
    public int countTriangles(int arr[]) {
        // code here
        int n=arr.length;
        if(n<3) return 0;
        int c=0;
        Arrays.sort(arr);
        for(int i=0;i<n-2;i++){
            int sideA=arr[i];
            for(int j=i+1;j<n-1;j++){
                int sideB=arr[j];
                int sum=sideA+sideB;
                int l=j+1,h=n-1,idx=-1;
                while(l<=h){
                    int mid=(l+h)>>1;
                    
                    if(arr[mid]<sum) {
                        l=mid+1;
                        idx=mid;
                    }
                    else h=mid-1;
                }
               if(idx!=-1) c+=idx-j;
                
            }
        }
        
        return c;
    }
}